Add class decorator support

himdel opened the pull request:
This makes it possible to run `decorate` not only on model instances, but also directly on the classes.

It also replaces Draper with a simpler solution.

Differences from Drapper:
  * there is only that `decorate` method, no attempt at implementing `decorator_class?` or any other Draper methods was made
  * that `decorate` method is also available on classes now
  * `delegate_all` is done automatically, but only for decorated instances, not much sense in using it for the decorated class
  * `decorate` returns `nil` when no decorator is found, instead of throwing

Usage:

`VmServer.decorate` will return `VmOrTemplateDecorator`
`vm_instance.decorate` will return an instance of `VmOrTemplateDecorator` which delegates to the original `vm_instance` object

adding

```
def self.fonticon
  "foobar"
end
```

to that decorator (note the `self`) will cause `VmServer.decorate.fonticon` to return `"foobar"`.

Closes ManageIQ/manageiq#10313

Done in preparation for https://github.com/ManageIQ/manageiq/issues/13199, cc @epwinchell 

~~TODO:~~:

   * [x] specs
   * [x] miq PR removing the draper dependency - https://github.com/ManageIQ/manageiq/pull/14044
   * [x] replace API ?decorators with supports - removing the last `.decorate` use from manageiq - https://github.com/ManageIQ/manageiq/pull/14040 (and https://github.com/ManageIQ/manageiq-ui-classic/pull/458)

---

Note that this *does* monkeypatch `ApplicationRecord` - adding a class and instance `decorate` method (and a `@_decorator` field to cache it once resolved).

To be able to decorate non-AR models, we'll need to `extend` and `include` the relevant mixin manually.

himdel commented:
@hayesr, @martinpovolny you were active in the original discussion about class decorators => ping :)

himdel commented:
Aand @skateman :)

himdel commented:
So..

`config/initializers/` is maybe not the best solution, but works (provided there is that `< ActiveRecord::Base`, otherwise it fails with `NoMethodError` trying to call `connection`).

Putting this in `app/models/` causes Draper's collection decorator take precedence, which doesn't work :(.

So.. I guess we can either keep it here, or remove draper and reimplement the decorate method here as well.

chrisarcand commented on lib/miq_decorator.rb:
```diff
@@ -0,0 +1,9 @@
+  def method_missing(*args)
+    @_decorated_instance.send(*args)
+  end
+end
```
I would strongly suggest using an actual delegator rather than crossing your fingers with method_missing.

Are you familiar with some of Ruby's core Delegator classes? I would use [`SimpleDelegator`](http://ruby-doc.org/stdlib-2.4.0/libdoc/delegate/rdoc/SimpleDelegator.html) here.

`class MiqDecorator < SimpleDelegator`

martinpovolny replied:
:+1:  on using the SimpleDelegator

chrisarcand commented on config/initializers/class_decorator.rb:
```diff
@@ -0,0 +1,22 @@
+  def decorate
+    self.class.decorator_for(self.class).new(self)
+  end
+end
```
(Follow this after addressing comment [here](https://github.com/ManageIQ/manageiq-ui-classic/pull/237/files#r97825145))

Patching something that doesn't belong to you is a bad idea, especially re-opening a class in our lovely project 😄 

Instead, you can do away with this entirely and create these methods in `MiqDecorator` itself. `MiqDecorator` can take in instances and klasses and return the proper decorator (which it owns) instance or class for you.

```ruby
class MiqDecorator < SimpleDelegator
  self << class
    def decorator_for(klass)
       "#{klass.name}Decorator".safe_constantize
    end

    def decorate(instance)
      decorator = instance_for(instance.class)
      decorator.new(instance)
    end
  end
end
```

Disclaimer: I just wrote this pseudo code up quick, a few details are plainly missing.

hayesr replied:
👍

Fryguy replied:
> Patching something that doesn't belong to you is a bad idea, especially re-opening a class in our lovely project 😄

I would take this a bit further and state that as you move to the REST API, you *can't* patch platform classes, so doing so here might make it harder to move forward.  However, using the decoration pattern above, it is clear which methods are owned by the UI and not the platform, and those can be converted to code that eventually decorates the result of the REST API.

chrisarcand commented on app/controllers/application_controller.rb:
```diff
@@ -959,7 +959,7 @@ def listicon_item(view, id = nil)
   # this always supersedes listicon_image if not nil
   def listicon_icon(item)
-    item.decorate.try(:fonticon) if item.decorator_class?
+    item.decorate.try(:fonticon)
```
With my other comments addressed, this now becomes a non-patched call, entirely dependent on the UI's own code, with `MiqDecorator.decorate(item)`. You of course can feel free to create a top level controller helper like `decorate_model` that does the same thing without `MiqDecorator` explicitly.

kbrock commented on spec/helpers/quadicon_helper_spec.rb:
```diff
@@ -963,7 +963,7 @@
       context "when item is not CIM or decorated" do
         before(:each) do
-          allow(item).to receive(:decorator_class?) { false }
+          allow(item).to receive(:decorate) { nil }
```
nit: could you use `and_return`?

```ruby
allow(item).to receive(:decorate).and_return(nil)
```

kbrock commented on config/initializers/class_decorator.rb:
```diff
@@ -0,0 +1,22 @@
+class ApplicationRecord < ActiveRecord::Base
+  class << self
+    def decorate
+      @_decorator || (@_decorator = decorator_for(self))
```
another nit. I think this is more clear with `||=`:

```ruby
@_decorator ||= decorator_for(self)
```

chrisarcand replied:
(But you won't need this code given proper delegation, anyway)

chrisarcand commented:
Your comment about initializers [here](https://github.com/ManageIQ/manageiq-ui-classic/pull/237#issuecomment-275117432) should be addressed/fixed given my feedback.

hayesr commented:
I think it might be possible to extend ApplicationRecord with an initializer block in `engine.rb`, but I think I like @chrisarcand's idea better. It feels a lot more object-oriented, and I bet we'll get some happy testing benefits as well.

Fryguy commented:
> Putting this in app/models/ causes Draper's collection decorator take precedence, which doesn't work :(.
> 
> So.. I guess we can either keep it here, or remove draper and reimplement the decorate method here as well.

It sounds like you want to remove Draper, so in that case, I would say remove draper.

chrisarcand commented:
Indeed, Draper *should* be replaced given where we are now.

Let me be clear that your implementation is understandable given that you were basically implementing how Draper does it. But Draper, with the UI split having been done, now doesn't make any sense for us at all because of how it patches things in much the same way. The UI doesn't really own it's models that it's trying to decorate anymore.

martinpovolny commented:
> But Draper, with the UI split having been done, now doesn't make any sense for us at all because of how it patches things in much the same way. The UI doesn't really own it's models that it's trying to decorate anymore.

Looking forward that's true. But for now it's not that much imporant. We need to be able to decorate to be able to do cleanups and move forward. 

As @Fryguy wrote, we might be decorating stuff returned from the API in the future. But for now we need the decorators for both instances and classes of the models.

himdel commented:
So... I've rebased and addressed most of the concerns, including moving the class resolution logic to `MiqDecorator`, rebased, etc.

**But**... do we really want to write `MiqDecorate.decorate_instance(instance)` and `MiqDecorate.decorate_class(VmOrTemplate)` instead of `instance.decorate` and `VmOrTemplate.decorate`? To me, that sounds like the exact opposite of right :).

himdel commented:
Tested this does not break the API, so we do *not* depend on ManageIQ/manageiq#14040.
Tested that removing Draper can be done after, so we do *not* depend on ManageIQ/manageiq#14044 either.

himdel commented:
... doing a resume of a gitter discussion - :point_up: [February 24, 2017 3:18 PM](https://gitter.im/ManageIQ/manageiq?at=58b04ed2f1a33b6275447193)

We will use `.decorate` for now, to keep the code same, and not have to depend on the API changes, 
**but** as step 2, we will stop patching ApplicationRecord and do `MiqDecorator.for(instance)` and `MiqDecorator.for(klass)` instead.

chrisarcand commented:
👍  Please note for anyone having an issue with the `MiqDecorator.for(object)` (which I think is great, personally...) You can use whatever helper method you want as long as it's within the scope of the UI project and doesn't directly edit ApplicationRecord or any other core class that doesn't belong to the UI (as we are doing above, simply to keep the `.decorate` syntax working). This is important for continued decoupling of the core and UI projects, making our lives betterer. Thanks. :)
